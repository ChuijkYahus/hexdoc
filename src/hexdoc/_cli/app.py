import code
import json
import logging
import os
import shutil
import sys
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from textwrap import dedent
from typing import Annotated, Union

from typer import Option, Typer

from hexdoc.core import ModResourceLoader
from hexdoc.minecraft import I18n
from hexdoc.minecraft.assets import AnimatedTexture

from . import render_block
from .utils.args import (
    DEFAULT_MERGE_DST,
    DEFAULT_MERGE_SRC,
    PathArgument,
    PropsOption,
    ReleaseOption,
    UpdateLatestOption,
    VerbosityOption,
)
from .utils.load import load_book, load_book_and_loader, load_books, load_common_data
from .utils.logging import repl_readfunc
from .utils.render import create_jinja_env, render_book
from .utils.sitemap import (
    assert_version_exists,
    delete_root_book,
    delete_updated_books,
    dump_sitemap,
    load_sitemap,
)

app = Typer(
    pretty_exceptions_enable=False,
    context_settings={
        "help_option_names": ["--help", "-h"],
    },
)
app.add_typer(render_block.app)


@app.command()
def list_langs(
    *,
    props_file: PropsOption,
    verbosity: VerbosityOption = 0,
):
    """Get the available language codes as a JSON list."""
    props, pm, _ = load_common_data(props_file, verbosity)
    with ModResourceLoader.load_all(
        props, pm, texture_render_dir=None, export=False
    ) as loader:
        langs = sorted(I18n.list_all(loader))
        print(json.dumps(langs))


@app.command()
def repl(
    *,
    props_file: PropsOption,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Start a Python shell with some helpful extra locals added from hexdoc."""
    props, pm, _ = load_common_data(props_file, verbosity)

    _, book, i18n, loader, context = load_book_and_loader(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
        export=False,
        texture_render_dir=Path("out/tmp/textures"),
    )

    repl_locals = dict(
        props=props,
        pm=pm,
        book=book,
        i18n=i18n,
        loader=loader,
        context=context,
    )

    code.interact(
        banner=dedent(
            f"""\
            [hexdoc repl] Python {sys.version}
            Locals: {', '.join(sorted(repl_locals.keys()))}"""
        ),
        readfunc=repl_readfunc(),
        local=repl_locals,
        exitmsg="",
    )


@app.command()
def export(
    *,
    props_file: PropsOption,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Run hexdoc, but skip rendering the web book - just export the book resources."""
    props, pm, _ = load_common_data(props_file, verbosity)
    load_book(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
        export=True,
        texture_render_dir=None,
    )


@app.command()
def render(
    output_dir: PathArgument = DEFAULT_MERGE_SRC,
    *,
    props_file: PropsOption,
    update_latest: UpdateLatestOption = True,
    release: ReleaseOption = False,
    clean: bool = False,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Export resources and render the web book."""

    # load data
    props, pm, version = load_common_data(props_file, verbosity)
    if not props.book:
        raise ValueError("Expected a value for props.book, got None")
    if not props.template:
        raise ValueError("Expected a value for props.template, got None")

    books, all_metadata = load_books(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
        texture_render_dir=output_dir / "textures",
    )

    logger = logging.getLogger(__name__)
    logger.info(f"update_latest={update_latest}, release={release}")

    # set up Jinja
    env = create_jinja_env(pm, props.template.include, props_file)

    template_names = (
        props.template.render
        if props.template.override_default_render
        else pm.default_rendered_templates(props.template.include)
    ) | props.template.extend_render

    templates = {
        Path(path): env.get_template(template_name)
        for path, template_name in template_names.items()
    }

    if not templates:
        raise RuntimeError(
            "No templates to render, check your props.template configuration "
            f"(in {props_file.as_posix()})"
        )

    if clean:
        shutil.rmtree(output_dir, ignore_errors=True)

    lang_names = {
        lang_: f"{i18n.localize('language.name')} ({i18n.localize('language.region')})"
        for lang_, (_, i18n, _) in books.items()
    }

    for should_render, version_, is_root in [
        (update_latest, "latest", False),
        (release, version, False),
        (update_latest and release, version, True),
    ]:
        if not should_render:
            continue
        for lang_, (book, i18n, context) in books.items():
            textures = context.png_textures
            animations = sorted(
                (
                    texture
                    for texture in textures.values()
                    if isinstance(texture, AnimatedTexture)
                ),
                key=lambda t: t.class_name,
            )

            render_book(
                props=props,
                pm=pm,
                lang=lang_,
                lang_names=lang_names,
                book=book,
                i18n=i18n,
                templates=templates,
                output_dir=output_dir,
                all_metadata=all_metadata,
                textures=textures,
                animations=animations,
                allow_missing=allow_missing,
                version=version_,
                is_root=is_root,
            )

    logger.info("Done.")


@app.command()
def merge(
    *,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    update_latest: UpdateLatestOption = True,
    release: ReleaseOption = False,
):
    # ensure at least the default language was built successfully
    if update_latest:
        assert_version_exists(root=src, version="latest")

    # TODO: figure out how to do this with pluggy (we don't have the props file here)
    # if is_release:
    #     assert_version_exists(src, GRADLE_VERSION)

    dst.mkdir(parents=True, exist_ok=True)

    # remove any stale data that we're about to replace
    delete_updated_books(src=src, dst=dst)
    if update_latest and release:
        delete_root_book(root=dst)

    # do the merge
    shutil.copytree(src=src, dst=dst, dirs_exist_ok=True)

    # rebuild the sitemap
    sitemap = load_sitemap(dst)
    dump_sitemap(dst, sitemap)


@app.command()
def serve(
    *,
    props_file: PropsOption,
    port: int = 8000,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    do_merge: Annotated[bool, Option("--merge/--no-merge")] = False,
    update_latest: bool = True,
    release: bool = True,  # you'd generally want --release for development
    clean: bool = False,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    book_root = dst if do_merge else src

    book_path = book_root.resolve().relative_to(Path.cwd())

    book_url = f"/{book_path.as_posix()}"

    os.environ |= {
        "DEBUG_GITHUBUSERCONTENT": "",
        "GITHUB_PAGES_URL": book_url,
    }

    print("Rendering...")
    render(
        props_file=props_file,
        output_dir=src,
        update_latest=update_latest,
        release=release,
        clean=clean,
        lang=lang,
        allow_missing=allow_missing,
        verbosity=verbosity,
    )

    if do_merge:
        print("Merging...")
        merge(
            src=src,
            dst=dst,
            update_latest=update_latest,
            release=release,
        )

    print(
        f"Serving web book at http://localhost:{port}{book_url} (press ctrl+c to exit)\n"
    )
    with HTTPServer(("", port), SimpleHTTPRequestHandler) as httpd:
        # ignore KeyboardInterrupt to stop Typer from printing "Aborted."
        # because it keeps printing after nodemon exits and breaking the output
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass


if __name__ == "__main__":
    app()
