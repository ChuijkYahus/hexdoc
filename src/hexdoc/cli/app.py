import code
import json
import logging
import os
import shutil
import sys
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from textwrap import dedent
from typing import Annotated, Union

from typer import Option, Typer

from hexdoc.core import ModResourceLoader
from hexdoc.minecraft import I18n, Tag
from hexdoc.minecraft.assets import (
    AnimatedTexture,
    HexdocAssetLoader,
    PNGTexture,
)
from hexdoc.utils.git import git_root

from . import render_block
from .utils.args import (
    DEFAULT_MERGE_DST,
    DEFAULT_MERGE_SRC,
    BranchOption,
    PathArgument,
    PropsOption,
    ReleaseOption,
    VerbosityOption,
)
from .utils.load import (
    load_book,
    load_common_data,
    render_textures_and_export_metadata,
)
from .utils.logging import repl_readfunc
from .utils.render import create_jinja_env, render_book
from .utils.sitemap import (
    assert_version_exists,
    delete_root_book,
    delete_updated_books,
    dump_sitemap,
    load_sitemap,
)

logger = logging.getLogger(__name__)

app = Typer(
    pretty_exceptions_enable=False,
    context_settings={
        "help_option_names": ["--help", "-h"],
    },
)
app.add_typer(render_block.app)


@app.command()
def list_langs(
    *,
    props_file: PropsOption,
    verbosity: VerbosityOption = 0,
):
    """Get the available language codes as a JSON list."""
    props, pm, _ = load_common_data(props_file, verbosity, "")
    with ModResourceLoader.load_all(
        props,
        pm,
        export=False,
    ) as loader:
        langs = sorted(I18n.list_all(loader))
        print(json.dumps(langs))


@app.command()
def repl(
    *,
    props_file: PropsOption,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Start a Python shell with some helpful extra locals added from hexdoc."""
    props, pm, _ = load_common_data(props_file, verbosity, "")

    loader = ModResourceLoader.load_all(
        props,
        pm,
        export=False,
    )

    _, book, i18n, context = load_book(
        props.book,
        props,
        pm,
        loader,
        lang,
        allow_missing,
    )

    repl_locals = dict(
        props=props,
        pm=pm,
        book=book,
        i18n=i18n,
        loader=loader,
        context=context,
    )

    code.interact(
        banner=dedent(
            f"""\
            [hexdoc repl] Python {sys.version}
            Locals: {', '.join(sorted(repl_locals.keys()))}"""
        ),
        readfunc=repl_readfunc(),
        local=repl_locals,
        exitmsg="",
    )


@app.command()
def export(
    output_dir: PathArgument = DEFAULT_MERGE_SRC,
    *,
    branch: BranchOption,
    release: ReleaseOption = False,
    props_file: PropsOption,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Run hexdoc, but skip rendering the web book - just export the book resources."""
    props, pm, plugin = load_common_data(props_file, verbosity, branch)

    loader = ModResourceLoader.clean_and_load_all(
        props,
        pm,
        export=True,
    )

    site_path = plugin.site_path(versioned=release)
    asset_loader = plugin.asset_loader(
        loader=loader,
        site_url=f"{loader.props.url}/{site_path.as_posix()}",  # TODO: urlencode
        asset_url=props.env.asset_url,
        render_dir=output_dir / site_path,
    )

    all_metadata = render_textures_and_export_metadata(loader, asset_loader)

    i18n = I18n.load_all(
        loader,
        allow_missing=allow_missing,
    )[props.default_lang]

    if props.book:
        load_book(
            book_id=props.book,
            pm=pm,
            loader=loader,
            i18n=i18n,
            all_metadata=all_metadata,
        )


@app.command()
def render(
    output_dir: PathArgument = DEFAULT_MERGE_SRC,
    *,
    branch: BranchOption,
    release: ReleaseOption = False,
    clean: bool = False,
    lang: Union[str, None] = None,
    props_file: PropsOption,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Export resources and render the web book."""

    # load data
    props, pm, plugin = load_common_data(props_file, verbosity, branch, book=True)

    if not props.book:
        raise ValueError("Expected a value for props.book, got None")
    if not props.template:
        raise ValueError("Expected a value for props.template, got None")

    site_path = plugin.site_book_path(
        lang=lang or props.default_lang,
        versioned=release,
    )

    books, all_metadata = load_books(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
    )

    # set up Jinja
    env = create_jinja_env(pm, props.template.include, props_file)

    template_names = (
        props.template.render
        if props.template.override_default_render
        else pm.default_rendered_templates(props.template.include)
    ) | props.template.extend_render

    templates = {
        Path(path): env.get_template(template_name)
        for path, template_name in template_names.items()
    }

    if not templates:
        raise RuntimeError(
            "No templates to render, check your props.template configuration "
            f"(in {props_file.as_posix()})"
        )

    if clean:
        shutil.rmtree(output_dir, ignore_errors=True)

    lang_names = {
        book_lang: i18n.localize_lang() for book_lang, (_, i18n, _) in books.items()
    }

    for book_lang, (book, i18n, context) in books.items():
        png_textures = PNGTexture.get_lookup(context.textures)
        animations = list(AnimatedTexture.get_lookup(context.textures).values())

        render_book(
            props=props,
            pm=pm,
            lang=book_lang,
            lang_names=lang_names,
            book=book,
            i18n=i18n,
            templates=templates,
            output_dir=output_dir,
            site_path=site_path,
            all_metadata=all_metadata,
            png_textures=png_textures,
            animations=animations,
            allow_missing=allow_missing,
            version=plugin.mod_version,
        )

    logger.info("Done.")


@app.command()
def merge(
    *,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    release: ReleaseOption = False,
):
    # ensure at least the default language was built successfully
    if update_latest:
        assert_version_exists(root=src, version="latest")

    # TODO: figure out how to do this with pluggy (we don't have the props file here)
    # if is_release:
    #     assert_version_exists(src, GRADLE_VERSION)

    dst.mkdir(parents=True, exist_ok=True)

    # remove any stale data that we're about to replace
    delete_updated_books(src=src, dst=dst)
    if update_latest and release:
        delete_root_book(root=dst)

    # do the merge
    shutil.copytree(src=src, dst=dst, dirs_exist_ok=True)

    # rebuild the sitemap
    sitemap = load_sitemap(dst)
    dump_sitemap(dst, sitemap)


@app.command()
def serve(
    *,
    props_file: PropsOption,
    port: int = 8000,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    branch: BranchOption,
    release: bool = True,  # generally want --release for development, looks nicer
    clean: bool = False,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    book_root = dst
    relative_root = book_root.resolve().relative_to(Path.cwd())

    base_url = f"http://localhost:{port}"
    book_url = f"{base_url}/{relative_root.as_posix()}"

    repo_root = git_root(props_file.parent)
    asset_root = repo_root.relative_to(Path.cwd())

    os.environ |= {
        # prepend a slash to the path so it can find the texture in the local repo
        # eg. http://localhost:8000/_site/src/docs/Common/...
        # vs. http://localhost:8000/Common/...
        "DEBUG_GITHUBUSERCONTENT": f"{base_url}/{asset_root.as_posix()}",
        "GITHUB_PAGES_URL": book_url,
    }

    print("Exporting...")
    export(
        branch=branch,
        release=release,
        props_file=props_file,
        allow_missing=allow_missing,
        verbosity=verbosity,
    )

    print("Rendering...")
    render(
        branch=branch,
        props_file=props_file,
        output_dir=src,
        release=release,
        clean=clean,
        lang=lang,
        allow_missing=allow_missing,
        verbosity=verbosity,
    )

    print("Merging...")
    merge(
        src=src,
        dst=dst,
        release=release,
    )

    print(f"Serving web book at {book_url} (press ctrl+c to exit)\n")
    with HTTPServer(("", port), SimpleHTTPRequestHandler) as httpd:
        # ignore KeyboardInterrupt to stop Typer from printing "Aborted."
        # because it keeps printing after nodemon exits and breaking the output
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass


if __name__ == "__main__":
    app()
