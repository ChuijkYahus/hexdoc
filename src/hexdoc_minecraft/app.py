import logging
from pathlib import Path
from zipfile import ZipFile

from github import Github
from hexdoc.cli.utils.args import (
    DEFAULT_BRANCH,
    PropsOption,
    ReleaseOption,
    VerbosityOption,
)
from hexdoc.cli.utils.load import export_metadata, load_common_data
from hexdoc.cli.utils.logging import setup_logging
from hexdoc.core.loader import ModResourceLoader
from hexdoc.minecraft import Tag
from typer import Typer

from .asset_loader import MinecraftAssetLoader
from .minecraft_assets import MinecraftAssetsRepo
from .piston_meta import VersionManifestV2

logger = logging.getLogger(__name__)

CACHE_ROOT = Path(".hexdoc_minecraft")

app = Typer(
    pretty_exceptions_enable=False,
    context_settings={
        "help_option_names": ["--help", "-h"],
    },
)


@app.command()
def fetch(
    version_id: str,
    *,
    verbosity: VerbosityOption = 1,
):
    """Download the Minecraft client jar."""
    setup_logging(verbosity)

    jar_path = CACHE_ROOT / version_id / "client.jar"

    manifest = VersionManifestV2.fetch()
    package = manifest.fetch_package(version_id)
    package.downloads.client.fetch_file(jar_path)

    logger.info("Done.")


@app.command()
def unzip(
    version_id: str,
    *,
    verbosity: VerbosityOption = 1,
):
    """Partially extract the Minecraft client jar."""
    setup_logging(verbosity)

    version_dir = CACHE_ROOT / version_id
    jar_path = version_dir / "client.jar"

    with ZipFile(jar_path) as jar:
        for name in jar.namelist():
            if name.startswith(
                (
                    "assets/minecraft/blockstates/",
                    "assets/minecraft/models/",
                    "assets/minecraft/textures/",
                )
            ):
                jar.extract(name, version_dir / "resources")

    logger.info("Done.")


@app.command()
def export(
    version_id: str,  # TODO: props.extra this
    *,
    assets_ref: str = "master",  # TODO: props.extra this
    branch: str = DEFAULT_BRANCH,
    release: ReleaseOption = False,
    props_file: PropsOption,
    verbosity: VerbosityOption = 0,
):
    """Export all textures."""
    props, pm, plugin = load_common_data(props_file, verbosity, branch)

    loader = ModResourceLoader.clean_and_load_all(
        props,
        pm,
        export=True,
    )

    asset_loader = MinecraftAssetLoader(
        loader=loader,
        asset_url=props.env.asset_url,
        gaslighting_items=Tag.GASLIGHTING_ITEMS.load(loader).value_ids_set,
        repo=MinecraftAssetsRepo(
            github=Github(),
            ref=assets_ref,
            version=version_id,
        ),
    )

    export_metadata(
        loader,
        asset_loader,
        site_path=plugin.site_path(versioned=release),
    )


if __name__ == "__main__":
    app()
