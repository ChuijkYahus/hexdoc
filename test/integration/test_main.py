# pyright: reportUnknownMemberType=false

import subprocess
from pathlib import Path
from typing import TYPE_CHECKING

import pytest
from hexdoc.cli.app import build, callback
from pytest import MonkeyPatch, TempPathFactory
from syrupy.assertion import SnapshotAssertion

from ..conftest import list_directory

# include: v/latest/main/assets
# include: v/latest/main/assets/hexcasting
# exclude: v/latest/main/assets/hexcasting/textures/block/akashic_ligature.png
EXCLUDE_GLOB = "**/assets/**/**"

CHECK_RENDERED_FILENAMES = [
    "v/latest/main/en_us/index.html",
    "v/latest/main/en_us/index.css",
    "v/latest/main/en_us/textures.css",
    "v/latest/main/en_us/index.js",
    "v/latest/main/en_us/hexcasting.js",
    "v/latest/main/en_us/.sitemap-marker.json",
]


@pytest.fixture(scope="session", autouse=True)
def call_callback():
    callback(quiet_lang=["ru_ru", "zh_cn"])


@pytest.fixture(scope="session", autouse=True)
def patch_versions(monkeysession: MonkeyPatch):
    # because pyright complains when we do this in the CI
    if TYPE_CHECKING:
        _hooks = None
    else:
        from hexdoc_hexcasting import _hooks

    monkeysession.setattr(_hooks, "GRADLE_VERSION", "MOD_VERSION")
    monkeysession.setattr(_hooks, "PY_VERSION", "PLUGIN_VERSION")
    monkeysession.setattr(_hooks, "FULL_VERSION", "FULL_VERSION")


@pytest.fixture(scope="session")
def app_output_dir(tmp_path_factory: TempPathFactory) -> Path:
    return tmp_path_factory.mktemp("app")


@pytest.fixture(scope="session")
def subprocess_output_dir(tmp_path_factory: TempPathFactory) -> Path:
    return tmp_path_factory.mktemp("subprocess")


@pytest.mark.hexcasting
@pytest.mark.dependency
def test_render_app_release(
    tmp_path_factory: TempPathFactory,
    snapshot: SnapshotAssertion,
    hexcasting_props_file: Path,
):
    app_output_dir = tmp_path_factory.mktemp("app")

    build(
        output_dir=app_output_dir,
        props_file=hexcasting_props_file,
        release=True,
        branch="main",
    )

    assert list_directory(app_output_dir, exclude_glob=EXCLUDE_GLOB) == snapshot


@pytest.mark.hexcasting
@pytest.mark.dependency
def test_render_app(app_output_dir: Path, hexcasting_props_file: Path):
    build(
        output_dir=app_output_dir,
        props_file=hexcasting_props_file,
        branch="main",
    )


@pytest.mark.hexcasting
@pytest.mark.dependency
def test_render_subprocess(subprocess_output_dir: Path, hexcasting_props_file: Path):
    cmd = [
        "hexdoc",
        "--quiet-lang=ru_ru",
        "--quiet-lang=zh_cn",
        "build",
        subprocess_output_dir.as_posix(),
        f"--props={hexcasting_props_file.as_posix()}",
        "--branch=main",
    ]
    subprocess.run(cmd, check=True)


@pytest.mark.hexcasting
@pytest.mark.dependency(depends=["test_render_app", "test_render_subprocess"])
def test_file_structure(
    app_output_dir: Path,
    subprocess_output_dir: Path,
    snapshot: SnapshotAssertion,
):
    app_list = list_directory(app_output_dir, exclude_glob=EXCLUDE_GLOB)
    subprocess_list = list_directory(subprocess_output_dir, exclude_glob=EXCLUDE_GLOB)

    assert app_list == subprocess_list
    assert app_list == snapshot


@pytest.mark.hexcasting
@pytest.mark.dependency(depends=["test_render_app", "test_render_subprocess"])
@pytest.mark.parametrize("filename", CHECK_RENDERED_FILENAMES)
def test_files(
    filename: str,
    app_output_dir: Path,
    subprocess_output_dir: Path,
    path_snapshot: SnapshotAssertion,
):
    app_file = app_output_dir / filename
    subprocess_file = subprocess_output_dir / filename

    assert app_file == path_snapshot

    # difficult to monkeypatch versions for subprocess, so this file will be different
    if not filename.endswith(".sitemap-marker.json"):
        assert app_file.read_bytes() == subprocess_file.read_bytes()
