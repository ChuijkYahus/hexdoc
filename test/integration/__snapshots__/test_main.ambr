# serializer version: 1
# name: test_file_structure
  list([
    'v',
    'v/latest',
    'v/latest/.sitemap-marker.json',
    'v/latest/apple-touch-icon-114x114.png',
    'v/latest/apple-touch-icon-120x120.png',
    'v/latest/apple-touch-icon-144x144.png',
    'v/latest/apple-touch-icon-152x152.png',
    'v/latest/apple-touch-icon-167x167.png',
    'v/latest/apple-touch-icon-180x180.png',
    'v/latest/apple-touch-icon-57x57.png',
    'v/latest/apple-touch-icon-60x60.png',
    'v/latest/apple-touch-icon-72x72.png',
    'v/latest/apple-touch-icon-76x76.png',
    'v/latest/favicon-16x16.png',
    'v/latest/favicon-180x180.png',
    'v/latest/favicon-196x196.png',
    'v/latest/favicon-32x32.png',
    'v/latest/favicon-64x64.png',
    'v/latest/favicon-96x96.png',
    'v/latest/favicon.ico',
    'v/latest/hexcasting.js',
    'v/latest/index.css',
    'v/latest/index.html',
    'v/latest/index.js',
    'v/latest/logo.png',
    'v/latest/mstile-270x270.png',
    'v/latest/mstile-310x150.png',
    'v/latest/mstile-310x310.png',
    'v/latest/mstile-70x70.png',
    'v/latest/textures.css',
  ])
# ---
# name: test_render_app_release
  list([
    'apple-touch-icon-114x114.png',
    'apple-touch-icon-120x120.png',
    'apple-touch-icon-144x144.png',
    'apple-touch-icon-152x152.png',
    'apple-touch-icon-167x167.png',
    'apple-touch-icon-180x180.png',
    'apple-touch-icon-57x57.png',
    'apple-touch-icon-60x60.png',
    'apple-touch-icon-72x72.png',
    'apple-touch-icon-76x76.png',
    'favicon-16x16.png',
    'favicon-180x180.png',
    'favicon-196x196.png',
    'favicon-32x32.png',
    'favicon-64x64.png',
    'favicon-96x96.png',
    'favicon.ico',
    'hexcasting.js',
    'index.css',
    'index.html',
    'index.js',
    'logo.png',
    'mstile-270x270.png',
    'mstile-310x150.png',
    'mstile-310x310.png',
    'mstile-70x70.png',
    'textures.css',
    'v',
    'v/VERSION',
    'v/VERSION/.sitemap-marker.json',
    'v/VERSION/apple-touch-icon-114x114.png',
    'v/VERSION/apple-touch-icon-120x120.png',
    'v/VERSION/apple-touch-icon-144x144.png',
    'v/VERSION/apple-touch-icon-152x152.png',
    'v/VERSION/apple-touch-icon-167x167.png',
    'v/VERSION/apple-touch-icon-180x180.png',
    'v/VERSION/apple-touch-icon-57x57.png',
    'v/VERSION/apple-touch-icon-60x60.png',
    'v/VERSION/apple-touch-icon-72x72.png',
    'v/VERSION/apple-touch-icon-76x76.png',
    'v/VERSION/favicon-16x16.png',
    'v/VERSION/favicon-180x180.png',
    'v/VERSION/favicon-196x196.png',
    'v/VERSION/favicon-32x32.png',
    'v/VERSION/favicon-64x64.png',
    'v/VERSION/favicon-96x96.png',
    'v/VERSION/favicon.ico',
    'v/VERSION/hexcasting.js',
    'v/VERSION/index.css',
    'v/VERSION/index.html',
    'v/VERSION/index.js',
    'v/VERSION/logo.png',
    'v/VERSION/mstile-270x270.png',
    'v/VERSION/mstile-310x150.png',
    'v/VERSION/mstile-310x310.png',
    'v/VERSION/mstile-70x70.png',
    'v/VERSION/textures.css',
    'v/latest',
    'v/latest/.sitemap-marker.json',
    'v/latest/apple-touch-icon-114x114.png',
    'v/latest/apple-touch-icon-120x120.png',
    'v/latest/apple-touch-icon-144x144.png',
    'v/latest/apple-touch-icon-152x152.png',
    'v/latest/apple-touch-icon-167x167.png',
    'v/latest/apple-touch-icon-180x180.png',
    'v/latest/apple-touch-icon-57x57.png',
    'v/latest/apple-touch-icon-60x60.png',
    'v/latest/apple-touch-icon-72x72.png',
    'v/latest/apple-touch-icon-76x76.png',
    'v/latest/favicon-16x16.png',
    'v/latest/favicon-180x180.png',
    'v/latest/favicon-196x196.png',
    'v/latest/favicon-32x32.png',
    'v/latest/favicon-64x64.png',
    'v/latest/favicon-96x96.png',
    'v/latest/favicon.ico',
    'v/latest/hexcasting.js',
    'v/latest/index.css',
    'v/latest/index.html',
    'v/latest/index.js',
    'v/latest/logo.png',
    'v/latest/mstile-270x270.png',
    'v/latest/mstile-310x150.png',
    'v/latest/mstile-310x310.png',
    'v/latest/mstile-70x70.png',
    'v/latest/textures.css',
  ])
# ---
